src-tauri/src/lib.rs

```diff
@@ -1,32 +1,123 @@
+use utils::http::HttpClientImpl;
 
 mod commands;
-mod http;
-mod note;
-mod service;
-mod user;
-
-use note::note_service::LocalNoteService;
-use note::note_service::RemoteNoteService;
-use note::NoteManager;
-use user::user_service::LocalUserService;
-use user::user_service::RemoteUserService;
-use user::UserManager;
+mod models;
+mod services;
+mod utils;
+
+#[async_trait]
+trait State: Send + Sync {
+    fn note_service(&self) -> &NoteService;
+    fn user_service(&self) -> &UserService;
+}
+
+struct LocalState {
+    note_service: LocalNoteService,
+    user_service: LocalUserService,
+}
+
+impl LocalState {
+    fn new(app: &mut App) -> Self {
+        let notes_store = app.handle().store_builder("notes.bin").build();
+        let note_service = LocalNoteService::new(notes_store);
+
+        let user_store = app.handle().store_builder("users.bin").build();
+        let user_service = LocalUserService::new(user_store);
+
+        Self {
+            note_service,
+            user_service,
+        }
+    }
+}
+
+#[async_trait]
+impl State for LocalState {
+    fn note_service(&self) -> &NoteService {
+        &self.note_service
+    }
+
+    fn user_service(&self) -> &UserService {
+        &self.user_service
+    }
+}
+
+struct RemoteState {
+    note_service: RemoteNoteService,
+    user_service: RemoteUserService,
+}
+
+impl RemoteState {
+    fn new() -> Self {
+        let remote_note_service =
+            RemoteNoteService::new("/notes".to_string(), HttpClientImpl::default());
+
+        let remote_user_service =
+            RemoteUserService::new("/users".to_string(), HttpClientImpl::default());
+
+        Self {
+            note_service: remote_note_service,
+            user_service: remote_user_service,
+        }
+    }
+}
+
+#[async_trait]
+impl State for RemoteState {
+    fn note_service(&self) -> &NoteService {
+        &self.note_service
+    }
+
+    fn user_service(&self) -> &UserService {
+        &self.user_service
+    }
+}
+
+enum NetworkMode {
+    Local,
+    Remote,
+}
+
+trait NetworkListener {
+    fn notify(&mut self, mode: NetworkMode);
+}
 
 struct AppState {
-    notes_manager: NoteManager,
-    users_manager: UserManager,
+    network_mode: NetworkMode,
+    local_service: LocalState,
+    remote_service: RemoteState,
 }
 
 impl AppState {
-    pub fn notes_manager(&self) -> &NoteManager {
-        &self.notes_manager
+    fn new(local: LocalState, remote: RemoteState) -> Self {
+        let local_state = local;
+        let remote_state = remote;
+
+        Self {
+            network_mode: NetworkMode::Local,
+            local_service: local_state,
+            remote_service: remote_state,
+        }
     }
 
-    pub fn users_manager(&self) -> &UserManager {
-        &self.users_manager
+    fn service(&self) -> &dyn State {
+        match self.network_mode {
+            NetworkMode::Local => &self.local_service,
+            NetworkMode::Remote => &self.remote_service,
+        }
+    }
+}
```

Comment:
Can't we just have LocalState and RemoteState as singletons, so they're not saved in the AppState? And then use memory address to pass as the current state?

Reply:
So in Rust the Singleton pattern is kinda "ugly" because it implies to use `unsafe`code.
Essentially, this code acts kind like a Singleton: the AppState instance is given as parameter in each of the commands, eg: `pub async fn users_index(app_handle: tauri::AppHandle) -> Result<Vec<User>, String>`.

The `service` method in the `State` struct returns a reference to the services, so there's always a unique instance of each service.